coil-gif/src/nonAndroidMain/kotlin/coil3/gif/AnimatedSkiaImage.kt

```diff
@@ -0,0 +1,233 @@
+
+        val isAnimationComplete =
+            maxIterationCount > 0 && totalElapsedTimeMs >= singleIterationDurationMs * maxIterationCount
+
+        if (isAnimationComplete) {
+            // The animation is complete, freeze on the last frame.
+            canvas.drawFrame(
+                frameIndex = codec.framesInfo.lastIndex,
+            )
+
+            if (!hasNotifiedAnimationEnd) {
+                onAnimationEnd?.invoke()
+                hasNotifiedAnimationEnd = true
+            }
+
+            return
+        }
+
+        val frameIndexToDraw = getFrameIndexToDraw(
+            frameDurationsMs = frameDurationsMs,
+            singleIterationDurationMs = singleIterationDurationMs,
+            totalElapsedTimeMs = totalElapsedTimeMs,
+        )
+
+        canvas.drawFrame(frameIndexToDraw)
+
+        if (!isAnimationComplete) {
+            invalidateTick++
+        }
+    }
+
+    /**
+     * Returns the index of the frame to draw based on the current time.
+     *
+     * If the animation is complete, the index of the last frame is returned.
+     *
+     * @param frameDurationsMs The list of frame durations in milliseconds.
+     */
+    private fun getFrameIndexToDraw(
+        frameDurationsMs: List<Int>,
+        singleIterationDurationMs: Int,
+        totalElapsedTimeMs: Long,
+    ): Int {
+        val currentIterationElapsedTimeMs = totalElapsedTimeMs % singleIterationDurationMs
+        return getFrameIndexToDrawWithinIteration(
+            frameDurationsMs = frameDurationsMs,
+            elapsedTimeMs = currentIterationElapsedTimeMs,
+        )
+    }
+
+    /**
+     * Returns the index of the frame to draw within the current iteration.
+     *
+     * @param frameDurationsMs The list of frame durations in milliseconds.
+     * @param elapsedTimeMs The elapsed time in milliseconds since the start of the iteration.
+     */
+    private fun getFrameIndexToDrawWithinIteration(
+        frameDurationsMs: List<Int>,
+        elapsedTimeMs: Long,
+    ): Int {
+        var accumulatedDuration = 0
+
+        for ((index, frameDuration) in frameDurationsMs.withIndex()) {
+            if (accumulatedDuration > elapsedTimeMs) {
+                return (index - 1).coerceAtLeast(0)
+            }
+
+            accumulatedDuration += frameDuration
+        }
+
+        return frameDurationsMs.lastIndex
+    }
+
+    private fun decodeFrame(frameIndex: Int): ByteArray {
+        if (tempBitmap.isClosed) {
+            throw IllegalStateException("Cannot decode frame: the bitmap is closed.")
+        }
+
+        codec.readPixels(tempBitmap, frameIndex)
+        return tempBitmap.readPixels(
+            dstInfo = codec.imageInfo,
+            dstRowBytes = codec.imageInfo.minRowBytes,
+        )!!
+    }
+
+    private fun Canvas.drawFrame(frameIndex: Int) {
+        animatedTransformation?.transform(this)
+
+        frames[frameIndex]?.let { frame ->
```

Comment:
We should probably block if the current frame we're trying to draw isn't available yet. We don't need to wait for the whole decode job to complete, however. We might need to spin in a while loop on JS to block.

Skipping drawing the current frame if it's not available yet could make it appear like the GIF is disappearing randomly on slower devices. It'll be up to the library user to set `bufferedFramesCount` high enough that we don't end up blocking.